src/gtimelog/email.py

```diff
@@ -0,0 +1,85 @@
+import collections
+import email
+import logging
+import smtplib
+from contextlib import closing
+from email.mime import text
+from email.utils import formataddr, parseaddr
+from gettext import gettext as _
+
+from gtimelog import DEBUG, __version__
+
+
+MailProtocol = collections.namedtuple('MailProtocol', 'factory, startssl')
+MAIL_PROTOCOLS = {
+    'SMTP': MailProtocol(smtplib.SMTP, False),
+    'SMTPS': MailProtocol(smtplib.SMTP_SSL, False),
+    'SMTP (StartTLS)': MailProtocol(smtplib.SMTP, True),
+}
+
+log = logging.getLogger('email')
+
+
+class EmailError(Exception):
+    pass
+
+
+def isascii(s):
+    return all(0 <= ord(c) <= 127 for c in s)
+
+
+def address_header(name_and_address):
+    if isascii(name_and_address):
+        return name_and_address
+    name, addr = parseaddr(name_and_address)
+    name = str(email.header.Header(name, 'UTF-8'))
+    return formataddr((name, addr))
+
+
+def subject_header(header):
+    if isascii(header):
+        return header
+    return email.header.Header(header, 'UTF-8')
+
+
+def prepare_message(sender, recipient, subject, body):
+    if isascii(body):
+        msg = text.MIMEText(body)
+    else:
+        msg = text.MIMEText(body, _charset="UTF-8")
+    if sender:
+        msg["From"] = address_header(sender)
+    msg["To"] = address_header(recipient)
+    msg["Subject"] = subject_header(subject)
+    msg["User-Agent"] = "gtimelog/{}".format(__version__)
+    return msg
+
+
+def send_email(protocol, server, port, username, sender, recipient, subject, body, password):
```

Comment:
This is an unwieldy set of parameters, and the ordering is a bit strange (password not next to the username).

There's a hidden design constraint that password needs to be the last parameter, because a partial application of this function is being passed as a callback somewhere.

We can do better than that, but I need to think a bit about how.